source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -16,22 +16,29 @@ namespace Http {
 DynamicFilterConfigProviderImpl::DynamicFilterConfigProviderImpl(
     FilterConfigSubscriptionSharedPtr&& subscription,
     const std::set<std::string>& require_type_urls,
-    Server::Configuration::FactoryContext& factory_context)
+    Server::Configuration::FactoryContext& factory_context,
+    FilterConfigProviderManagerImpl& filter_config_provider_manager, const std::string& provider_id)
     : subscription_(std::move(subscription)), require_type_urls_(require_type_urls),
-      tls_(factory_context.threadLocal()),
+      tls_(factory_context.threadLocal()), provider_id_(provider_id),
+      filter_config_provider_manager_(filter_config_provider_manager),
       init_target_("DynamicFilterConfigProviderImpl", [this]() {
         subscription_->start();
         // This init target is used to activate the subscription but not wait
         // for a response. It is used whenever a default config is provided to be
         // used while waiting for a response.
         init_target_.ready();
       }) {
-  subscription_->filter_config_providers_.insert(this);
+  // It should be 1:1 mapping due to shared dynamic filter config provider.
+  ASSERT(!subscription_->filter_config_provider_.has_value());
+  subscription_->filter_config_provider_.emplace(this);
   tls_.set([](Event::Dispatcher&) { return std::make_shared<ThreadLocalConfig>(); });
 }
 
 DynamicFilterConfigProviderImpl::~DynamicFilterConfigProviderImpl() {
-  subscription_->filter_config_providers_.erase(this);
+  ASSERT(subscription_->filter_config_provider_.has_value());
+  subscription_->filter_config_provider_.reset();
+  // Remove the provider from the provider manager.
+  filter_config_provider_manager_.dynamic_filter_config_providers_.erase(provider_id_);
```

Comment:
Given my recent work on callbacks I'm a little paranoid about logic like this. Is it guaranteed that the provider will be destroyed before the manager? Does the manger itself need to be locked to make sure it is still alive? I don't have my head fully wrapped around the lifetime of all of the different objects here. If it is guaranteed safe can you add more lifetime comments here and elsewhere?

Reply:
I followed RDS code to pass in a provider manager reference here, which I thought would be safe. I tried to track down the ownership of provider and provider manager and looks like both of them were owned by filter factory lambda: https://github.com/envoyproxy/envoy/blob/2509ee6159883d75f86a3af57fe0e6cc4b74bbdc/source/extensions/filters/network/http_connection_manager/config.cc#L167

Where `singletons` keeps a shared pointer of the provider manager, and `filter_config` keeps a unique pointer of the provider. I think the same lifecycle applies to RDS provider and provider manager. IIUC, this means that provider and manager destruction order is not deterministic? If so, to resolve this, we'd probably pass in the config provider manager shared pointer all the way through `HttpConnectionManagerConfig` to provider, instead of passing in a reference?

Reply:
> I think the good start is that adding a ASSERT(subscriptions_.empty()); in ~FilterConfigProviderManagerImpl

I added an assert.